### PR TITLE
Add agentic workflows that keep Microsoft.Extensions.AI aligned with OpenTelemetry GenAI semantic-conventions

### PR DESCRIPTION
**Pull request by jeffhandley**
### Summary

`Microsoft.Extensions.AI` emits OpenTelemetry GenAI spans/metrics, and the upstream [`open-telemetry/semantic-conventions-genai`](https://github.com/open-telemetry/semantic-conventions-genai) spec evolves continuously. Keeping our instrumentation current is manual and easy to miss.

This PR adds a pair of **GitHub Agentic Workflows** (built on [`gh-aw`](https://github.com/github/gh-aw)) that continuously scan the upstream conventions and maintain a **single evergreen draft PR** proposing the compensating changes to `Microsoft.Extensions.AI`. A human still reviews and merges -- the automation only ever produces a *draft* PR and never touches anything outside `src/Libraries/Microsoft.Extensions.AI*`, `test/Libraries/Microsoft.Extensions.AI*`, and `docs/`.

It adds **only CI/automation assets** (`.github/**`) -- no product code or public API changes.

### How it works

The system follows the **OrchestratorOps** pattern (one deterministic orchestrator, one scoped agentic worker):

- **Orchestrator** (`meai-otel-genai-orchestrator.yml`) -- a plain, non-agentic workflow. On a daily schedule (or manual dispatch), it resolves the upstream ref to a concrete commit SHA and hands the worker a single, fully-specified target. It invokes the worker as a **reusable workflow (`workflow_call`)** so the run keeps the orchestrator's actor and passes gh-aw's activation role check.
- **Worker** (`meai-otel-genai-worker.md`) -- the agentic workflow. A deterministic host step (`meai-otel-genai-worker-setup.sh`) classifies the maintained PR (fresh / incremental / advisory / adopt / no-op) and reads prior state from a machine-readable block in the PR body. The agent then applies the **`update-otel-genai-conventions` skill** to implement/refresh the integration, builds + tests it, and maintains the PR through gh-aw **safe outputs** (create / push / update / comment / mark-ready / no-op).

Safety model: read-only agent job; all writes go through **allow-listed safe outputs**; **threat-detection** screens the untrusted upstream content before anything is published; a **PAT pool** avoids single-token rate limits; the maintained PR stays a **draft** requiring human review.

Lifecycle handled: first-run creation, daily incremental updates (append-only commits), behind-non-draft advisory (comment only), adoption of a human-started PR, reviewer-feedback incorporation, and release mark-ready.

### What is included

| Area | Files |
|---|---|
| Orchestrator + worker | `meai-otel-genai-orchestrator.yml`, `meai-otel-genai-worker.md` (+ compiled `.lock.yml`) |
| Deterministic host scripts | `meai-otel-genai-orchestrator-discover.sh`, `meai-otel-genai-worker-setup.sh` |
| Domain skill | `.github/skills/update-otel-genai-conventions/**` |
| gh-aw scaffolding | `.github/aw/`, `.github/mcp.json`, `.github/agents/`, `copilot-setup-steps.yml` |
| PAT pool | `.github/workflows/shared/pat_pool.*`, `validate-pat-pool.yml` |

17 files, `.github/**` only.

### Testing

**In brief:** Validated end-to-end on a dedicated simulation repository by "time-traveling" through pinned upstream commits. The full lifecycle -- clean-state creation and an incremental update -- runs **green**, including a real restore/build/test and threat-detection. The simulation also **uncovered and fixed two workflow bugs**, after which every reachable scenario passed.

**Details:**

- **Method** -- Because the workflow reacts to upstream state, it was exercised against a standalone test repo by forcing `upstream_ref` to specific `semantic-conventions-genai` commits (T0 baseline, T1 a later commit), driving the orchestrator -> worker chain end-to-end on real GitHub Actions runners.

- **Fresh, from a clean state (@ T0):** all jobs green (discover -> worker activation -> agent -> **threat-detection** -> safe outputs). A draft PR was created on the evergreen branch with `automation`+`area-ai` labels, a **real code change** (`document` modality in `OtelMessageSerializer.cs` + tests), a clean **build across net8.0/net9.0/net10.0/netstandard2.0 (0 warnings)** with tests passing, and the machine-readable tracking block.

- **Incremental update (@ T1):** the worker correctly detected the PR was behind, integrated the delta (`gen_ai.request.reasoning.level`), and **appended a second commit** -- the original commit was preserved (no force-push, no amend), and the tracking state advanced. This validates the "every update is an incremental commit" guarantee.

- **Bugs found and fixed during testing** (folded into their originating commits):
  1. On the fresh path the worker committed while checked out on the base branch, so the new PR branch matched its base and produced an **empty patch** -- fixed by detaching `HEAD` before committing.
  2. `create-pull-request: allow-empty` caused gh-aw to **skip patch generation entirely**, shipping a content-free PR and failing threat-detection -- removed it and emit `no-op` when a scan yields nothing to integrate.

- **Deterministic host scripts** (`*-discover.sh`, `*-setup.sh`) were separately exercised with a chaos/property/exhaustive test harness (classification, state parsing, release gating, wake-gate, idempotence), which caught and fixed additional edge cases in the setup logic.

### Notes / follow-ups

- Threat-detection currently authenticates via a coalesced PAT-pool secret to work around a gh-aw issue where the detection job does not inherit `needs: pat_pool`.
- The maintained PR is always a **draft**; nothing merges without human review.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/extensions/pull/7611)

### CODE REVIEW

File: .github/workflows/meai-otel-genai-orchestrator.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    permissions:
+      contents: write
+      pull-requests: write
+      issues: write
+      # The worker's compiled reusable workflow requests discussions: write on its
+      # gh-aw-generated jobs (the compiler grants the union of all safe-output scopes,
+      # even ones we don't use). A reusable workflow can't exceed the caller's
+      # permissions, so this must be granted here even though we never touch discussions.
+      discussions: write
+      actions: read
```
[the remark this answers is not in the document]
Comment (jeffhandley) [reply .github/workflows/meai-otel-genai-orchestrator.yml:82]: The comment explains why this is here.

File: .github/workflows/shared/pat_pool.README.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        needs.pat_pool.outputs.pat_number == '1', secrets.COPILOT_PAT_1,
+        needs.pat_pool.outputs.pat_number == '2', secrets.COPILOT_PAT_2,
+        needs.pat_pool.outputs.pat_number == '3', secrets.COPILOT_PAT_3,
+        needs.pat_pool.outputs.pat_number == '4', secrets.COPILOT_PAT_4,
+        needs.pat_pool.outputs.pat_number == '5', secrets.COPILOT_PAT_5,
+        needs.pat_pool.outputs.pat_number == '6', secrets.COPILOT_PAT_6,
+        needs.pat_pool.outputs.pat_number == '7', secrets.COPILOT_PAT_7,
+        needs.pat_pool.outputs.pat_number == '8', secrets.COPILOT_PAT_8,
+        needs.pat_pool.outputs.pat_number == '9', secrets.COPILOT_PAT_9,
+        'NO COPILOT PAT AVAILABLE')
```
[the remark this answers is not in the document]
Comment (jeffhandley) [reply .github/workflows/shared/pat_pool.README.md:109]: A workflow _can_ have a fallback but doesn't _need_ to employ that.

File: .github/workflows/meai-otel-genai-worker.md [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+        needs.pat_pool.outputs.pat_number == '1', secrets.COPILOT_PAT_1,
+        needs.pat_pool.outputs.pat_number == '2', secrets.COPILOT_PAT_2,
+        needs.pat_pool.outputs.pat_number == '3', secrets.COPILOT_PAT_3,
+        needs.pat_pool.outputs.pat_number == '4', secrets.COPILOT_PAT_4,
+        needs.pat_pool.outputs.pat_number == '5', secrets.COPILOT_PAT_5,
+        needs.pat_pool.outputs.pat_number == '6', secrets.COPILOT_PAT_6,
+        needs.pat_pool.outputs.pat_number == '7', secrets.COPILOT_PAT_7,
+        needs.pat_pool.outputs.pat_number == '8', secrets.COPILOT_PAT_8,
+        needs.pat_pool.outputs.pat_number == '9', secrets.COPILOT_PAT_9,
+        'NO COPILOT PAT AVAILABLE')
```
[the remark this answers is not in the document]
Comment (jeffhandley) [reply .github/workflows/meai-otel-genai-worker.md]: Once again, Copilot, it says workflows _can_ fall back. We aren't.

File: .github/workflows/meai-otel-genai-worker.md [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+  create-pull-request:
+    draft: true
+    labels: [automation, area-ai]
+    base-branch: main
+    preserve-branch-name: true
+    if-no-changes: "warn"
+    allowed-files:
+      - "src/Libraries/Microsoft.Extensions.AI*/**"
+      - "test/Libraries/Microsoft.Extensions.AI*/**"
+      - "docs/**"
```
Comment (tannergooding) [inline .github/workflows/meai-otel-genai-worker.md]: Does this need to be any docs in the repo or should it also be filtered down to just the MEAI folders?

File: .github/workflows/meai-otel-genai-worker.md [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+points at the current commit without switching to it is safe (and required -- see
+below). So:
+
+- **Fresh path** (merged-but-behind, closed, or no PR): the run is checked out **on the
+  base branch** (`main`), so committing now would advance `main` itself and leave the PR
+  branch you create pointing at the **same commit as its base** -- the generated patch is
+  then **empty** (and, with threat-detection enabled, the run hard-fails because no
+  patch/bundle is produced for the detection job to screen). **First detach `HEAD` at the
+  checked-out commit** so your commit advances a detached `HEAD` and leaves `main` where it
+  is: `git checkout --detach`. Detaching at the current commit is safe -- it does not switch
```
Comment (tannergooding) [inline .github/workflows/meai-otel-genai-worker.md]: Some agents don't always follow commands like this accurately/reliably. Given that this is essentially one of the bugs you had previously found and are trying to handle, it should probably be done in a host-script so its deterministic (like in dotnet/core#10469)

File: .github/workflows/meai-otel-genai-worker.md [hunk 6]
```diff
@@ -0,0 +1,754 @@
...
+  # agent integrates. It runs in a separate gh-aw job that authenticates via the coalescing
+  # token below.
+  threat-detection:
+    engine:
+      id: copilot
+      env:
+        # Workaround for github/gh-aw#43917: the detection job's `needs` omit `pat_pool`, so the
+        # main engine's case(needs.pat_pool...) token can't resolve here. Authenticate by
+        # coalescing the pool PAT secrets directly -- the first non-empty one wins.
+        COPILOT_GITHUB_TOKEN: ${{ secrets.COPILOT_PAT_0 || secrets.COPILOT_PAT_1 || secrets.COPILOT_PAT_2 || secrets.COPILOT_PAT_3 || secrets.COPILOT_PAT_4 || secrets.COPILOT_PAT_5 || secrets.COPILOT_PAT_6 || secrets.COPILOT_PAT_7 || secrets.COPILOT_PAT_8 || secrets.COPILOT_PAT_9 || secrets.COPILOT_GITHUB_TOKEN }}
```
Comment (tannergooding) [inline .github/workflows/meai-otel-genai-worker.md]: What ensures we actually distribute across PATs? Assuming that they all are normally defined, we're going to weight everything behind PAT_0 every time and not actually load balance, right?

File: .github/scripts/meai-otel-genai-worker-setup.sh [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+		if [ -z "$ref" ]; then
+			db="$(gh api "repos/${UPSTREAM_REPO}" -q '.default_branch' 2>/dev/null || true)"
+			[ -n "$db" ] && sha="$(gh api "repos/${UPSTREAM_REPO}/commits/${db}" -q '.sha' 2>/dev/null || true)"
+		else
+			sha="$(gh api "repos/${UPSTREAM_REPO}/commits/${ref}" -q '.sha' 2>/dev/null || true)"
+		fi
+		[ -n "$sha" ] && { printf '%s' "$sha"; return 0; }
+		[ "$attempt" -lt 3 ] && sleep 2
+	done
+	return 0
```
Comment (tannergooding) [inline .github/scripts/meai-otel-genai-worker-setup.sh]: We always return 0 here (success) even on failure?

**Review by tannergooding — APPROVED** [review]
Overall LGTM. A couple minor questions/nits.

A local Copilot review didn't find anything that I hadn't spotted either.

**Review by jeffhandley — COMMENTED** [review]
Thanks for the review. Summary of the changes: narrowed and then removed the docs scope, moved the fresh-path HEAD detach into the host setup script, made the detection job select its Copilot PAT from the pool by the same random selection as the agent job (enabled by recompiling on gh-aw v0.82.6), and made `resolve_sha` fail honestly instead of returning success. Per-thread responses follow.

File: .github/workflows/meai-otel-genai-worker.md (hunk 4 above)
Comment (jeffhandley) [reply .github/workflows/meai-otel-genai-worker.md]: Filtered down, then removed entirely. I first narrowed the `create-pull-request` and `push-to-pull-request-branch` allow-lists to just `src/Libraries/Microsoft.Extensions.AI*/**` and `test/Libraries/Microsoft.Extensions.AI*/**`, dropping the broad `docs/**`. On a closer look there are no docs for the agent to update beyond API doc-comments in the source (already covered by those src paths): the library `README.md` files carry no gen-ai convention references, and gh-aw protects `README.md` by name regardless. So I also removed the "update any affected docs" instruction from the prompt.

File: .github/workflows/meai-otel-genai-worker.md (hunk 5 above)
Comment (jeffhandley) [reply .github/workflows/meai-otel-genai-worker.md]: Agreed -- moved into a host script. The fresh-path HEAD detach now runs in `meai-otel-genai-worker-setup.sh` (the deterministic "Set up the run context" step) before the agent, instead of asking the agent to run `git checkout --detach`. Verified in a simulation run: the setup step logged `Detached HEAD at <sha>`, the fresh-path commit landed on the PR branch, and `main` did not advance.

File: .github/workflows/meai-otel-genai-worker.md (hunk 6 above)
Comment (jeffhandley) [reply .github/workflows/meai-otel-genai-worker.md]: Good catch -- and this line is fixed now too. Two jobs pick the token:

- **Agent job:** the `pat_pool` job collects only the populated `COPILOT_PAT_*` entries and selects one by random index (`PAT_INDEX=$(( RANDOM % count ))`); the agent authenticates via `case(needs.pat_pool.outputs.pat_number == 'N', secrets.COPILOT_PAT_N)`. So it is distributed by random selection, not weighted behind PAT_0.
- **Detection job (this line):** it previously coalesced the secrets directly (first-non-empty, so PAT_0) because it could not consume the `pat_pool` output (github/gh-aw#43917). Recompiling on gh-aw v0.82.6 lets the detection job depend on `pat_pool`, so it now uses the same `case(needs.pat_pool.outputs.pat_number, ...)` selection. Both jobs load-balance across the populated PATs now.

File: .github/scripts/meai-otel-genai-worker-setup.sh (hunk 7 above)
Comment (jeffhandley) [reply .github/scripts/meai-otel-genai-worker-setup.sh]: Fixed. `resolve_sha` now returns non-zero when it exhausts its retries without resolving a SHA. The call site runs it in a command substitution guarded with `|| true` so `set -euo pipefail` does not abort before the existing explicit empty-result check, which emits the clear diagnostic and exits. (The separate `resolve_release_commit` keeps `return 0` on no-tag, since "no release tag" is a legitimate soft-success there.)

**Review by tannergooding — APPROVED** [review]
